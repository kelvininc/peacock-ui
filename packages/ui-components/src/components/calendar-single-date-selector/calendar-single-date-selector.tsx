import { Component, Event, EventEmitter, h, Host, Prop } from '@stencil/core';
import { formatDate, fromISO, isDateSame } from '../../utils/date.helper';
import { IClickDateEvent } from '../calendar/calendar.types';
import { ICalendarSingleDateSelector, ICalendarSingleDateSelectorEvents, ISelectDate } from './calendar-single-date-selector.types';

@Component({
	tag: 'kv-calendar-single-date-selector',
	shadow: true
})
export class KvCalendarSingleDateSelector implements ICalendarSingleDateSelector, ICalendarSingleDateSelectorEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) minDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) maxDate?: string;

	/** @inheritdoc */
	@Event() selectDate: EventEmitter<ISelectDate>;

	private onClickDate = ({ detail: { event, payload: date } }: CustomEvent<IClickDateEvent>): void => {
		const clickedDateMoment = fromISO(date);

		// check if clicked date is the same as the selected date
		if (this.selectedDate !== undefined && isDateSame(clickedDateMoment, this.selectedDate)) {
			this.selectDate.emit({ event });
			return;
		}

		this.selectDate.emit({ event, payload: formatDate(clickedDateMoment) });
	};

	public getSelectedDates = (): [string] | undefined => {
		if (this.selectedDate) {
			return [this.selectedDate];
		}
	};

	render() {
		return (
			<Host>
				<kv-calendar
					initialDate={this.initialDate ?? this.selectedDate}
					onClickDate={this.onClickDate}
					selectedDates={this.getSelectedDates()}
					disabledDates={this.disabledDates}
					minDate={this.minDate}
					maxDate={this.maxDate}
					exportparts="calendar-container,month-container,day-container"
				/>
			</Host>
		);
	}
}
