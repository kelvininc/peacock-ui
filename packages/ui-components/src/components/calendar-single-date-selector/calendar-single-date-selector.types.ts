import { EventEmitter } from '@stencil/core';
import { ICalendar } from '../calendar/calendar.types';

export interface ICalendarSingleDateSelector extends Omit<ICalendar, 'selectedDates'> {
	/** (optional) Currently selected date */
	selectedDate?: string;
}

export interface ICalendarSingleDateSelectorEvents {
	/** Emitted when day is selected */
	selectDate: EventEmitter<ISelectDate>;
}

export interface ISelectDate {
	event: MouseEvent;
	payload?: string;
}
