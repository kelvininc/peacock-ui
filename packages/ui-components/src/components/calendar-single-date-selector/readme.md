# _kv-calendar-single-date-selector_

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-calendar-single-date-selector></kv-calendar-single-date-selector>

<!-- With selected date -->
<kv-calendar-single-date-selector selectedDate="2022-08-04"></kv-calendar-single-date-selector>

<!-- With initial date -->
<kv-calendar-single-date-selector initialDate="2022-08-04"></kv-calendar-single-date-selector>

<!-- With disabled dates-->
<kv-calendar-single-date-selector [disabledDates]="['2021-12-01', '2021-12-04', '2021-12-05']"></kv-calendar-single-date-selector>

<!-- With min and max date -->
<kv-calendar-single-date-selector minDate="2021-12-04" maxDate="2023-02-12"></kv-calendar-single-date-selector>
```


### React

```tsx
import React from 'react';

import { KvCalendarSingleDateSelector } from '@kelvininc/react-ui-components';

export const KvCalendarSingleDateSelectorExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvCalendarSingleDateSelector />

		{/*-- With selected date --*/}
		<KvCalendarSingleDateSelector selectedDate="2022-08-04" />

		{/*-- With initial date --*/}
		<KvCalendarSingleDateSelector initialDate="2021-12-03" />

		{/*-- With disabled dates --*/}
		<KvCalendarSingleDateSelector disabledDates={['2021-12-01', '2021-12-04', '2021-12-05']} />

		{/*-- With min and max --*/}
		<KvCalendarSingleDateSelector minDate="2021-12-04" maxDate="2023-02-12" />
	</>
);
```



## Properties

| Property        | Attribute       | Description                        | Type       | Default     |
| --------------- | --------------- | ---------------------------------- | ---------- | ----------- |
| `disabledDates` | --              | (options) The disabled dates array | `string[]` | `[]`        |
| `initialDate`   | `initial-date`  | (optional) Initial date            | `string`   | `undefined` |
| `maxDate`       | `max-date`      | (optional) Maximum accepted date   | `string`   | `undefined` |
| `minDate`       | `min-date`      | (optional) Minimum accepted date   | `string`   | `undefined` |
| `selectedDate`  | `selected-date` | (optional) Currently selected date | `string`   | `undefined` |


## Events

| Event        | Description                  | Type                       |
| ------------ | ---------------------------- | -------------------------- |
| `selectDate` | Emitted when day is selected | `CustomEvent<ISelectDate>` |


## Dependencies

### Used by

 - [kv-single-date-select-dropdown](../single-date-select-dropdown)

### Depends on

- [kv-calendar](../calendar)

### Graph
```mermaid
graph TD;
  kv-calendar-single-date-selector --> kv-calendar
  kv-calendar --> kv-action-button-icon
  kv-calendar --> kv-calendar-day
  kv-action-button-icon --> kv-action-button
  kv-action-button-icon --> kv-icon
  kv-action-button-icon --> kv-badge
  kv-single-date-select-dropdown --> kv-calendar-single-date-selector
  style kv-calendar-single-date-selector fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


