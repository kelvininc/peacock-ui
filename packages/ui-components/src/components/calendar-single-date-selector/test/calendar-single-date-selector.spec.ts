import { SpecPage } from '@stencil/core/internal';
import { KvCalendarSingleDateSelector } from '../calendar-single-date-selector';
import { newSpecPage } from '@stencil/core/testing';

describe('Calendar Single Date Selector (unit tests)', () => {
	let page: SpecPage;
	let component: KvCalendarSingleDateSelector;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvCalendarSingleDateSelector],
				html: '<kv-calendar-single-date-selector></kv-calendar-single-date-selector>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when `getSelectedDates` is called', () => {
		describe('and there is no selected date', () => {
			let result: [string] | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarSingleDateSelector],
					html: '<kv-calendar-single-date-selector></kv-calendar-single-date-selector>'
				});
				component = page.rootInstance;
				result = component.getSelectedDates();
			});

			it('should return undefined', () => {
				expect(result).toBeUndefined();
			});
		});

		describe('and there is a selected date', () => {
			let result: [string] | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarSingleDateSelector],
					html: '<kv-calendar-single-date-selector selected-date="2022-04-01"></kv-calendar-single-date-selector>'
				});
				component = page.rootInstance;
				result = component.getSelectedDates();
			});

			it('should return an array with the selected date', () => {
				expect(result).toEqual(['2022-04-01']);
			});
		});
	});
});
