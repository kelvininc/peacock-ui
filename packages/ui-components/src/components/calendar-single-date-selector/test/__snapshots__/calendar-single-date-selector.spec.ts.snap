// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Single Date Selector (unit tests) when uses default props should match the snapshot 1`] = `
<kv-calendar-single-date-selector>
  <mock:shadow-root>
    <kv-calendar exportparts="calendar-container,month-container,day-container"></kv-calendar>
  </mock:shadow-root>
</kv-calendar-single-date-selector>
`;
